packages/mui-material/src/Avatar/Avatar.js

```diff
@@ -163,17 +163,23 @@ const Avatar = React.forwardRef(function Avatar(inProps, ref) {
 
   let children = null;
 
-  // Use a hook instead of onError on the img element to support server-side rendering.
-  const loaded = useLoaded({ ...imgProps, src, srcSet });
-  const hasImg = src || srcSet;
-  const hasImgNotFailing = hasImg && loaded !== 'error';
-
   const ownerState = {
     ...props,
-    colorDefault: !hasImgNotFailing,
     component,
     variant,
   };
+
+  // Use a hook instead of onError on the img element to support server-side rendering.
+  const loaded = useLoaded({
+    ...imgProps,
+    ...(typeof slotProps.img === 'function' ? slotProps.img(ownerState) : slotProps.img),
```

Comment:
Off topic:

I think this common pattern (`...(typeof slotProps.x === 'function' ? slotProps.x(ownerState) : slotProps.x)`) could have a util in the future, it would also be useful for users: https://github.com/mui/material-ui/pull/44350#issuecomment-2501804154

What do you think?

Reply:
Could be. It is already handled by the `useSlot` but I think this is an edge case.